Builder circuit breaker

tbenr opened the pull request:
introduces `BuilderCircuitBreaker` interface and implementation

moves all the fallback check responsibility into `ExecutionLayerManagerImpl` (entirely removed from `BlockOperationSelectorFactory`)

We have three new params:
`--Xbuilder-circuit-breaker-enabled` -> default `true`
`--Xbuilder-circuit-breaker-window` -> default `32`
`--Xbuilder-circuit-breaker-allowed-faults` -> default `8`

fixes #6058

## Documentation

- [ ] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [ ] I thought about adding a changelog entry, and added one if I deemed necessary.

tbenr committed:
initial circuit breaker draft

rolfyone commented on beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/BlockOperationSelectorFactory.java:
```diff
@@ -149,14 +150,31 @@ public Consumer<BeaconBlockBodyBuilder> createSelector(
+                        Optional.of(BuilderForcedFallbackReason.TRANSITION_NOT_FINALIZED);
+                  } else {
+                    final int window = 32;
+                    final int missingBlockThreshold = 10;
```
assuming a couple of CLI args here with sensible limits will feed these two values...

tbenr committed:
removes fallback conditions checks from BlockOperationSelectorFactory
dedicated BuilderCircuitBreaker for better modularity\testing

tbenr committed:
fixUT

tbenr committed:
add CLI params
implemented tests

tbenr committed:
revert dirty println

tbenr committed:
cleanup BuilderForcedFallbackReason

tbenr committed:
fix ut

StefanBratanov reviewed (approved):
LGTM :)

tbenr committed:
fix last block header counting
move to chain builder to prepare the state
pass BuilderCircuitBreaker to Stub too

tbenr committed:
fix dependency rule

tbenr commented on build.gradle:
```diff
@@ -94,20 +94,15 @@ dependencyRules {
-          ":storage"
-      ]
-    }
+    ['statetransition', 'executionlayer', 'weaksubjectivity'].forEach( {
```
@ajsutton I added 'executionlayer' in the list of the bad guys since I needed to use the `InMemoryStorageSystemBuilder` for unit testing. Any concerns?

tbenr committed:
improve variable naming

tbenr committed:
fix blsVerifyDeposit reset in RecentChainDataTest

tbenr committed:
added change log

tbenr committed:
UT:
 remove storage dependency
 apply state transition via spec

tbenr committed:
UT:
 remove storage dependency
 apply state transition via spec

tbenr committed:
fix ut

rolfyone commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/ExecutionLayerManagerStub.java:
```diff
@@ -14,25 +14,43 @@
+  @Override
+  public SafeFuture<ExecutionPayloadHeader> builderGetHeader(
+      ExecutionPayloadContext executionPayloadContext, BeaconState state) {
+    LOG.info("Builder Circuit Breaker isEngaged: " + builderCircuitBreaker.isEngaged(state));
```
do we really want to see this a lot?

tbenr replied:
we shouldn't, it's when we activate the stub. So it is voluntarily verbose.

tbenr committed:
improve exception catching
